### DISCUSSION TITLE
Table 特性 AutoGenerateClass 和 AutoGenerateColumn 优先级疑问

### DISCUSSION THREAD

**Discussion by densen2014** [dotnetcore/BootstrapBlazor#3762] [Q&A] [answered]
为了少写代码, 我整个类写了 AutoGenerateClass: Searchable = true, 但是某一字段我又不想给Searchable,于是写了AutoGenerateColumn: Searchable = false, 渲染出来这个列还是使用了AutoGenerateClass指定的特性, 应该怎么处理才合理呢?

```
[AutoGenerateClass(Searchable = true, Filterable = true, Sortable = true, ShowTips = true)]
public class Sysprocesses  
{
    [AutoGenerateColumn(Searchable = false,Sortable =false ,Filterable =false )]
    public string? Country { get; set; }
}
```

表格渲染写上
```
    <TableColumns>
        <TableColumn TType="string" TItem="Sysprocesses"  @bind-Field="@context.Country" Searchable="false" />
    </TableColumns>
```
貌似也没用

**Comment by ArgoZhang** [comment]
@densen2014 我来复现一下

**Reply by ArgoZhang** [reply]
@densen2014 找到原因了。因为 `AutoGenerateClass` `AutoGenerateColumn` 这类标签不允许使用特殊数据类型如 `bool?` 所以这里只能使用基本的 `bool` 数据类型，这就导致只能使用 **继承** 的概念，即父级优先，但是我们可以更改 `ITableColumn` 把这个接口的 `Searchable` 更改为 `bool?` 类型，未设置时是 `null` 通过这个我们代码就可以判断

- 如果 `ITableColumn` 有值即用户设定值，覆盖所有标签设置。
- 如果仅设置 `AutoGenerateClass` `AutoGenerateColumn` 则应该 `AutoGenerateColumn` 优先

这个逻辑可否接受？以此类推好多属性都需要改动如 `Filterable`

请测试 `fix-table-column` 分支代码，已更新

个人理解，如果写了 `AutoGenerateClass` 又写了 `AutoGenerateColumn` 我个人更倾向是合并属性值，但是如果 `Class` 设置了 `Searchable=\"true\"` `Column` 标签 `Searchable` 默认值为 `false` 有时候更希望最终值为 `true` 但是目前无法识别 `Column` 是否使用了默认值，还是用户自己设置了 `Searchable=\"false\"`

**Reply by densen2014** [reply]
需要一个清晰的优先级指引, 我个人认为应该是 TableColumn >AutoGenerateColumn > AutoGenerateClass , 最优先使用最终手K特性, 符合懒癌患者.

**Comment by densen2014** [comment]
测试结果 
1. `<TableColumn Searchable="false" Sortable="false" />` 能覆盖  `[AutoGenerateClass(Searchable = true, Sortable = true)]`
2. `[AutoGenerateColumn(Searchable = false,Sortable =false)]` 能覆盖  `[AutoGenerateClass(Searchable = true, Sortable = true)]`
3. 改造后 AutoGenerateColumn 初始化会往离奇的方向走, 只设置 [AutoGenerateColumn(Searchable =true)] , 就不走  [AutoGenerateClass(Searchable = true, Filterable = true, Sortable = true, ShowTips = true)] 逻辑了对吗. 只要字段设置了 AutoGenerateColumn 某一个特性, 就完全没应用 AutoGenerateColumn 的设置了

**Comment by densen2014** [comment]
上一版修改后. 继承 class 标签中设置的参数值, 当source.Sortable为false 并且 dest.Sortable = null , 不设置dest.Sortable = source.Sortable, 最终还是空. 应该为
 ```
if (source.Sortable || dest.Sortable == null) dest.Sortable = source.Sortable;
```
才对吧


----------

测试代码

```
   [Fact]
    public void CopyValue_Sortable_Ok()
    {
        ITableColumn? tc;
        AutoGenerateColumnAttribute? columnAttribute = null;
        AutoGenerateClassAttribute? classAttribute = null;

        //测试自动生成特性
        //columnAttribute = new AutoGenerateColumnAttribute()
        //{
        //    Sortable = true
        //};

        //测试类特性
        classAttribute = new AutoGenerateClassAttribute()
        {
            Sortable = false,
        };

        if (columnAttribute == null)
        {
            // 未设置 AutoGenerateColumnAttribute 时使用默认值
            tc = new InternalTableColumn("Name", typeof(string));

            if (classAttribute != null)
            {
                // AutoGenerateClassAttribute 设置时继承类标签值
                tc.InheritValue(classAttribute);
                Assert.Null(tc.Sortable);
            }
            else
            {
                //都没设置,使用默认值 null
                Assert.Null(tc.Sortable);
            }
        }
        else
        {
            // 设置 AutoGenerateColumnAttribute 时合并 AutoGenerateClassAttribute 参数值

            tc = columnAttribute;
            Assert.True(tc.Sortable);

            if (classAttribute != null)
            {
                // AutoGenerateClassAttribute 设置时继承类标签值
                columnAttribute.InheritValue(classAttribute);
            }

            tc = columnAttribute;
            Assert.True(tc.Sortable);

        }

        if (columnAttribute != null)
        {
            //实体-列
            var attrInterface = (ITableColumn)columnAttribute;
            attrInterface.Sortable= true;

            // 替换属性 手写优先
            tc.CopyValue(attrInterface);

            Assert.True(tc.Sortable);
        }

    }
```

**Reply by ArgoZhang** [reply]
拉一下最新代码，我合并了那个继承逻辑，这单元测试是可以通过的

**Reply by ArgoZhang** [reply, marked as answer]
@densen2014 结案总结

- 只写了 `AutoGenerateClassAttribute` 以其参数值为准
- 只写了 `AutoGenerateColumnAttribute` 以其参数值为准
- 写了 `AutoGenerateClassAttribute` 并且写了 `AutoGenerateColumnAttribute` 合并两者参数值取其默认值的相反值
举例 `Filterable` 其默认值为 `false`
`AutoGenerateClassAttribute` 设置 `Filterable="true"` `AutoGenerateColumnAttribute` 设置 `Filterable="false"` 最终 UI 取值 `true`
`AutoGenerateClassAttribute` 设置 `Filterable="false"` `AutoGenerateColumnAttribute` 设置 `Filterable="true"` 最终 UI 取值 `true`

- 写了 `TableColumn` 参数值情况如下
最新更新 `TableColumn` 的 `Filterable` 更改为可为空数据类型，其默认值由原来的 `false` 更改为 `null`，可由此判断，只要 `TableColumn` 的参数 `Filterable` 值不为空，即以 `TableColumn` 值为最终 UI 值
